ADR 8 - RIO monad in cardano cli (#60)
input-output-hk/cardano-node-wiki: docs/ADR-8-Use-RIO-in-cardano‐cli.md

  ByronCommand cmds ->
    firstExceptT ByronClientError $ runByronClientCommand cmds
  CompatibleCommands cmd ->
    firstExceptT (BackwardCompatibleError (renderAnyCompatibleCommand cmd)) $
      runAnyCompatibleCommand cmd
...
```
- As a result we have a lot of errors wrapped in errors which makes the code unwieldly and difficult to compose with other code blocks. See image below of incidences of poor composability where we use `firstExceptT` (and sometimes `first`) to wrap errors in other errors.

![image](https://github.com/user-attachments/assets/6a50e8f2-9140-4b7e-afa5-ff778d696742)
 
- The ExceptT IO is a known anti-pattern for these reasons and others as per:
  - https://github.com/haskell-effectful/effectful/blob/master/transformers.md
  - https://tech.fpcomplete.com/blog/2016/11/exceptions-best-practices-haskell/

# Proposed Solution

Use the RIO monad which is based on the [ReaderT design pattern](https://tech.fpcomplete.com/blog/2017/06/readert-design-pattern/).

### Pros 
I opt for the `RIO env a` monad instead of `ReaderT e m a` because: 
- We get additional [logging functionality](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#g:5)

> - Explicit environment dependencies e.g `logError :: HasLogFunc env => Text -> RIO env ()`

Reviewer: I don't understand this point. In `ReaderT e m a` you have explicit environment dependencies as well. The example doesn't fit here?

> - Better composability i.e no more errors that wrap errors (see above).

Reviewer: `ReaderT` does not provide error handling so this point seems to be not related?

### Cons
- `RIO` is hardcoded to IO so we cannot add additional transformer layers e.g `RIO Env (StateT Int IO) a`
- Implicit error flow. Errors are thrown via `IO` exceptions.

### Exception handling

#### Exception type
```haskell
data CustomCliException where
  CustomCliException
    :: (Show error, Typeable error, Error error, HasCallStack)
    => error -> CustomCliException

deriving instance Show CustomCliException

instance Exception CustomCliException where
  displayException (CustomCliException e) =
    unlines
      [ show (prettyError e)
      , prettyCallStack callStack
      ]
```

#### Top level handler 
```haskell
executeRio :: RIO () () -> IO ()
executeRio r = do
  runRIO () r
    `catch` ( \e ->
                case fromException e of

>                   Just custom@(CustomCliException{}) -> putStrLn $ displayException custom

Reviewer: 1. [ADR 7](https://github.com/input-output-hk/cardano-node-wiki/wiki/ADR-7-CLI-Output-Presentation) says we should log to stderr
2. I'm not quite seeing the point of differentiation between two cases: `CustomCliException` and the other one. You're doing `displayException` in either case.

Reviewer: >This ensures we capture all exceptions.

Catching `SomeException` would do the same here.

>We handle CustomCliException explicitly for structured, user-facing errors, while the Nothing case acts as a safety net for unexpected exceptions.

Again, this separation is not shown anywhere in the code. A next person reading this ADR will have no clue what is the role of `CustomCliException`.

Author: > Catching SomeException would do the same here.

Are you saying to explicitly pattern match on this? I don't understand your point here.

Reviewer: >Are you saying to explicitly pattern match on this? I don't understand your point here.
Yes, `SomeException` is a catch all for all exceptions (asynchronous as well).

Author: How is this different that the `Nothing` case? It's going to handle all other exceptions.

Reviewer: Exactly, it is not - so what's specific about `CustomCliException` to pattern match it in the handler?

Author: https://github.com/input-output-hk/cardano-node-wiki/pull/60#discussion_r1943309730

                  Nothing -> putStrLn $ displayException e
            )
```

### Exception Handling Mechanism: Pros & Cons

#### Pros

1. **Unified Exception Type**
```haskell

> data CustomCliException where

Reviewer: CustomCliException is defined two times, differently in this ADR. Which version is correct?

  CustomCliException :: (Show e, Typeable e, Error e, HasCallStack) => e -> CustomCliException

deriving instance Show CustomCliException

instance Exception CustomCliException where
  displayException (CustomCliException e) =
    unlines
      [ show (prettyError e)
      , prettyCallStack callStack
      ]
```

> - Simplifies Top-Level Handling: All errors are caught as `CustomCliException`.

Reviewer: This ADR does not explain how this type should be used. You can't `fromException` other exceptions into `CustomCliExceptions`, so it will be useful only when explicitly thrown.

Author: >  You can't fromException other exceptions into CustomCliExceptions

I don't understand this objection. Can you clarify?

> Is this a temporary solution to bridge between ExceptT errors and exceptions, or a final design?

This is a _suggestion_ of a final design.

The purpose of `CustomCliException` is to represent explicitly thrown, structured errors that are meaningful to our application. For other exceptions (e.g., runtime errors or unexpected failures), we rely on the `Nothing` case in our handler.

Reviewer: >I don't understand this objection. Can you clarify?

You've answered my question in the next paragraph of your comment. 👍🏻 

>This is a suggestion of a final design.

>The purpose of CustomCliException is to represent explicitly thrown, structured errors that are meaningful to our application. For other exceptions (e.g., runtime errors or unexpected failures), we rely on the Nothing case in our handler.

Ok. The thing is, that explanation should be in the ADR itself, so that it's clear for the next person reading it.

Author: It will be a strict rule when the ADR gets merged. I.e I want the team's buy in.

Reviewer: >All errors are caught as CustomCliException.

This is nowhere demonstrated.

Author: ![image](https://github.com/user-attachments/assets/d8dac19d-5653-4470-81f8-b1b56f130a3e)

Reviewer: https://github.com/input-output-hk/cardano-node-wiki/pull/60#discussion_r1943239307

- Consistent Reporting: Ensures all errors are formatted uniformly via `prettyError`.

2. **CallStack Inclusion**

- Embeds CallStack to trace error origins, aiding debugging.

3. **Polymorphic Error Support**

- Flexibility: Wraps any error type so long at the required instances exist.

#### Cons
1. **Type Erasure**

> - Loss of Specificity: Existential quantification erases concrete error types, preventing pattern-matching on specific errors.

Reviewer: ```suggestion
- Loss of Specificity: Existential quantification erases concrete error types, preventing pattern-matching on specific errors. That may make specific error recovery logic harder to implement.
```

# Decision

> - Error handling: All errors will be converted to exceptions that will be caught by a single exception handler at the top level.

Reviewer: Converted how? This is not explained in the examples.

Author: https://github.com/input-output-hk/cardano-node-wiki/pull/60/files#diff-c83af4b681074eae4a9a579fb6a393539599918fd6e7c9e069d8f2422590b48aR100

Reviewer: Linking to a whole PR from an ADR is not ideal:
1. That PR is too big to illustrate how it should be used. Now a reader has to go through that PR, sift through non-relevant parts, read it, and figure out what's the idea here.
2. A PR until it's merged is a living thing, can still change.
3. There's a potential bitrot possibility in case yet another github organisation migration happens, github.com dissapearance etc.

Could you provide a self-contained example what do you mean here? 

Additionally that PR does not show any errors' conversion.

Do you mean using exceptions as a replacement for `Except e m a` here?

# Status
- [ ] Adopted YYYY/MM/DD

# Required Reading

- [ReaderT design pattern](https://tech.fpcomplete.com/blog/2017/06/readert-design-pattern/).
- [RIO](https://tech.fpcomplete.com/haskell/library/rio/)

# Context

> In `cardano-cli` we are using `ExceptT someError IO someResult` pattern 292 times in function types in our codebase for different error types.
> ![image](https://github.com/user-attachments/assets/f405dfca-7d75-404c-b291-3d5140fc8093)

Reviewer: ```suggestion
In `cardano-cli` we are using `ExceptT someError IO someResult` pattern 292 times in function types in our codebase for different error types.

![image](https://github.com/user-attachments/assets/f405dfca-7d75-404c-b291-3d5140fc8093)
```

The vast majority of these errors are not used to recover, they are propagated and reported to the user.

> ```

Reviewer: ```suggestion
```haskell
```

### Pros 
- Additional [logging functionality](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#g:5)
- Explicit environment dependencies e.g `logError :: HasLogFunc env => Text -> RIO env ()`
- Better composability i.e no more errors that wrap errors (see above).

### Cons
- `RIO` is hardcoded to IO so we cannot add additional transformer layers e.g `RIO Env (StateT Int IO) a`
- Implicit error flow. Errors are thrown via GHC exceptions in `IO`. See exception handling below.

## Exception handling

### Exception type
```haskell
data CustomCliException where
  CustomCliException
    :: (Show error, Typeable error, Error error, HasCallStack)
    => error -> CustomCliException

deriving instance Show CustomCliException

instance Exception CustomCliException where
  displayException (CustomCliException e) =
    unlines
      [ show (prettyError e)
      , prettyCallStack callStack
      ]
```

The purpose of `CustomCliException` is to represent explicitly thrown, structured errors that are meaningful to our application. 
For other exceptions (e.g., runtime errors or unexpected failures), we rely on the `Nothing` case in our proposed top level

> handler below:

Reviewer: Why the handler needs to pattern match exceptions?

Author: 1. We need a callstack and the `Error error` constraint enforces a pretty rendering for the exception.
2. Fair point it is also of the type `SomeException`.

Reviewer: >1. We need a callstack and the `Error error` constraint enforces a pretty rendering for the exception.

I see. But that will only give you a call stack up to the place where the `CustomCliException` was used to wrap our current error type.

As an alternative I was thinking about this solution:
1. Adding `HasCallStack` to each of our error constructors
2. Add  hand-written `Exception` instance to render callstack
```haskell
-- | Our standard error clas
class Error e where
  prettyError :: e -> Text

-- | our typical CLI error type
data CmdError
  = HasCallStack => CmdOtherError Text

-- | Needs standalone deriving instance
deriving instance Show CmdError

-- | already existing Error instance
instance Error CmdError where
  prettyError (CmdOtherError e) = "CmdError: " <> e

-- | Needs manually written Exception instance to print callstack
instance Exception CmdError where
  displayException ce@(CmdOtherError _) = 
    T.unpack (prettyError ce) <> "\n" <> fromString (prettyCallStack callStack)
```

The downside is that it requires a lot more changes. Eventually we can get to that point, but as an interim solution your `CustomCliException` is fine here. 👍🏻

Author: > which is a bit simpler and still conveys the same error information.

I'm happy to go with this. The idea of the `CustomCliException` type was to make it slightly more explicit so that it stands out more when we happen to forget an `Error` instance or a `HasCallStack` constraint. However we are big boys and what you proposed works for me.

### Top level handler 
```haskell

> executeRio :: RIO () () -> IO ()

Reviewer 2: I would hide the `env` parameter of the `RIO` now that it is easy, unless we are sure it is going to remain `()`, and even if we are. Maybe with a type synonym. Something like: `type CIO = RIO ()`. That way if we eventually want to add an environment we won't have to change every type signature

Author: Something like `type EmptyEnvironment = ()`? We will eventually use it in logging.

Reviewer: I wouldn't bother with hiding it. I think we can just write function types for any environment e.g. `RIO e a`. 

I'm wondering if maybe we could put stuff like `LocalNodeConnectInfo` for online commands, or environment variables into environment. So an online command using environment variables would have type like for example:
```haskell
doSomeCliCommand 
  :: Has LocalNodeConnectInfo e
  => Has EnvCli e
  => HasCallStack
  => Foo -> Bar -> RIO e ()

class Has a e where
  obtain :: a -> e
```
But that's out of scope for this ADR.

Besides that, maybe it be useful to have a type synonym providing `HasCallStack` everywhere? e.g.

```haskell
-- CIO, like Cardano IO 
type CIO e a = HasCallStack => RIO e a
```

Author: > I'm wondering if maybe we could put stuff like LocalNodeConnectInfo for online commands, or environment variables into environment.

100%. That is the point of the `Has` classes :+1: .

> besides that, maybe it be useful to have a type synonym providing HasCallStack everywhere?

Good suggestion, yes we should do this.

Reviewer 2: Ah, so we have traits. That works too

executeRio r = do
  runRIO () r
    `catch` ( \e ->
                case fromException e of
                  Just custom@(CustomCliException{}) -> hPutStrLn stderr $ displayException custom
                  Nothing -> hPutStrLn stderr $ displayException e
            )
```

### Usage Example of CustomCliException

https://github.com/IntersectMBO/cardano-cli/pull/1033

### Exception Handling Mechanism: Pros & Cons

#### Pros

1. **Unified Exception Type**

- Simplifies Top-Level Handling: All errors are caught as `CustomCliException`.
- Consistent Reporting: Ensures all errors are formatted uniformly via `prettyError`.

2. **CallStack Inclusion**

- Embeds CallStack to trace error origins, aiding debugging.

3. **Polymorphic Error Support**

> - Flexibility: Wraps any error type so long at the required instances exist.

Reviewer 2: ```suggestion
- Flexibility: Wraps any error type so long as the required instances exist.
```

#### Cons
1. **Type Erasure**

- Loss of Specificity: Existential quantification erases concrete error types, preventing pattern-matching on specific errors. That may make specific error recovery logic harder to implement.

# Decision
- Error handling: All errors will be converted to exceptions that will be caught by a single exception handler at the top level.
- Top level monad: [RIO](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#t:RIO). 

# Consequences
- This should dramatically improve our code's composability and remove many unnecessary error types. 

> - Readability concerning what errors can be thrown will be negatively impacted. However `ExceptT` already lies about what exceptions can be thrown because it is not limited to the error type stated in `ExceptT`'s type signature. In other words `IO` can implicitly throw other `Exception`s. 

Reviewer 2: ```suggestion
- Readability concerning what errors can be thrown will be negatively impacted. However, `ExceptT` already lies about what exceptions can be thrown because it is not limited to the error type stated in `ExceptT`'s type signature. In other words, `IO` can implicitly throw other `Exception`s. 
```

> - Initially this will be adopted under the "compatible" group of commands so `cardano-cli` will have a design split temporarily. Once we are happy with the result we will propagate to the rest of `cardano-cli`

Reviewer 2: Just from the point of the document structure. I notice there are essentially three pros and cons sections. One in `Proposed Solution`, one in `Exception Handling Mechanism:`, and one disguised as `Consequences`. Would it make sense to unify them?

> - Initially this will be adopted under the "compatible" group of commands so `cardano-cli` will have a design split temporarily. Once we are happy with the result we will propagate to the rest of `cardano-cli`

Reviewer 2: ```suggestion
- Initially, this will be adopted under the "compatible" group of commands so `cardano-cli` will have a design split temporarily. Once we are happy with the result we will propagate to the rest of `cardano-cli`
```

  = TransactionWriteFileError !(FileError ())

runClientCommandTransactions
  :: ()
  => ClientCommandTransactions
  -> ExceptT ExampleTransactionCmdError IO ()
runClientCommandTransactions DummyClientCommandToRun =
  ...
  left $
    TransactionWriteFileError $
      FileError "dummy.file" ()
``` 

Proposed change:

```haskell
data ClientCommandTransactions = DummyClientCommandToRun

data ExampleClientCommand = ClientCommandTransactions ClientCommandTransactions

topLevelRunCommand :: ExampleClientCommand -> RIO () ()
topLevelRunCommand (ClientCommandTransactions txsCmd) =
  runClientCommandTransactions txsCmd

runClientCommandTransactions
  :: HasCallStack
  => ClientCommandTransactions
  -> RIO () ()
runClientCommandTransactions DummyClientCommandToRun =
  ...
  throwIO $
    CustomCliException $
      FileError "dummy.file" ()
```

We have eliminated `data ExampleClientCommandErrors` and `data CmdError` and improved the composability of our code.

### Pros 
- Additional [logging functionality](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#g:5)
- Explicit environment dependencies e.g `logError :: HasLogFunc env => Text -> RIO env ()`

> - Better composability i.e no more errors that wrap errors (see above).

Reviewer 3: ```suggestion
- Better composability i.e no more errors that wrap errors (see above).
- Less bookeping required: no more wrapping low-level errors into high-level errors
```

Author: I think the point above already captures this

deriving instance Show CustomCliException

instance Exception CustomCliException where
  displayException (CustomCliException e) =
    unlines
      [ show (prettyError e)
      , prettyCallStack callStack
      ]
```

The purpose of `CustomCliException` is to represent explicitly thrown, structured errors that are meaningful to our application. 

### Exception Handling Mechanism: Pros & Cons

#### Pros

1. **Unified Exception Type**

- Simplifies Top-Level Handling: All errors are caught as `CustomCliException`.
- Consistent Reporting: Ensures all errors are formatted uniformly via `prettyError`.

2. **CallStack Inclusion**

- Embeds CallStack to trace error origins, aiding debugging.

3. **Polymorphic Error Support**

- Flexibility: Wraps any error type so long as the required instances exist.

#### Cons
1. **Type Erasure**

- Loss of Specificity: Existential quantification erases concrete error types, preventing pattern-matching on specific errors. That may make specific error recovery logic harder to implement.

# Decision
- Error handling: All errors will be converted to exceptions that will be caught by a single exception handler at the top level.

> - Top level monad: [RIO](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#t:RIO). 

Reviewer 3: ```suggestion
- Top level monad: [RIO](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#t:RIO).
- We agree not to catch `CustomCliException` except in the top-level handler. If the need for catching an exception arises, we locally use an `Either` or `ExceptT` pattern instead.
- `CustomCliException` should only be thrown within the `RIO` monad. Pure code is still not allowed to throw exceptions.
```

# Status
- [ ] Adopted YYYY/MM/DD

# Context

> - In `cardano-cli` We have the (anti-)pattern `ExceptT someError IO someResult` 292 times in our codebase.

Reviewer: ```suggestion
In `cardano-cli` we are using `ExceptT someError IO someResult` pattern 292 times in function types in our codebase for different error types.
```

![image](https://github.com/user-attachments/assets/f405dfca-7d75-404c-b291-3d5140fc8093)

> - The vast majority of these errors are not used to recover, they are propagated and reported to the user.

Reviewer: ```suggestion
The vast majority of these errors are not used to recover, they are propagated and reported to the user.
```

```

> main :: IO ()
> main = toplevelExceptionHandler $ do
>   Crypto.cryptoInit
> 
>   envCli <- getEnvCli
> 
>   GHC.mkTextEncoding "UTF-8" >>= GHC.setLocaleEncoding
> #ifdef UNIX
>   _ <- setFileCreationMask (otherModes `unionFileModes` groupModes)
> #endif
>   co <- Opt.customExecParser pref (opts envCli)
> 
>   orDie (docToText . renderClientCommandError) $ runClientCommand co
> ...
> ...
> ...
> 
> runClientCommand :: ClientCommand -> ExceptT ClientCommandErrors IO ()
> runClientCommand = \case
>   AnyEraCommand cmds ->
>     firstExceptT (CmdError (renderAnyEraCommand cmds)) $ runAnyEraCommand cmds
>   AddressCommand cmds ->
>     firstExceptT AddressCmdError $ runAddressCmds cmds
>   NodeCommands cmds ->
>     runNodeCmds cmds
>       & firstExceptT NodeCmdError
>   ByronCommand cmds ->
>     firstExceptT ByronClientError $ runByronClientCommand cmds
>   CompatibleCommands cmd ->
>     firstExceptT (BackwardCompatibleError (renderAnyCompatibleCommand cmd)) $
>       runAnyCompatibleCommand cmd
>   HashCmds cmds ->
>     firstExceptT HashCmdError $ runHashCmds cmds
>   KeyCommands cmds ->
>     firstExceptT KeyCmdError $ runKeyCmds cmds
>   LegacyCmds cmds ->
>     firstExceptT (CmdError (renderLegacyCommand cmds)) $ runLegacyCmds cmds
>   QueryCommands cmds ->
>     firstExceptT QueryCmdError $ runQueryCmds cmds
>   CliPingCommand cmds ->
>     firstExceptT PingClientError $ runPingCmd cmds
>   CliDebugCmds cmds ->
>     firstExceptT DebugCmdError $ runDebugCmds cmds
>   Help pprefs allParserInfo ->
>     runHelp pprefs allParserInfo
>   DisplayVersion ->
>     runDisplayVersion

Reviewer: This example is too long, and it takes some time to understand it, like why there's  `CPP` used. How about shortening it? 

```suggestion
main :: IO ()
main = toplevelExceptionHandler $ do

  envCli <- getEnvCli
  co <- Opt.customExecParser pref (opts envCli)
  orDie (docToText . renderClientCommandError) $ runClientCommand co
  
...

runClientCommand :: ClientCommand -> ExceptT ClientCommandErrors IO ()
runClientCommand = \case
  AnyEraCommand cmds ->
    firstExceptT (CmdError (renderAnyEraCommand cmds)) $ runAnyEraCommand cmds
  AddressCommand cmds ->
    firstExceptT AddressCmdError $ runAddressCmds cmds
  NodeCommands cmds ->
    runNodeCmds cmds
      & firstExceptT NodeCmdError
  ByronCommand cmds ->
    firstExceptT ByronClientError $ runByronClientCommand cmds
  CompatibleCommands cmd ->
    firstExceptT (BackwardCompatibleError (renderAnyCompatibleCommand cmd)) $
      runAnyCompatibleCommand cmd
...
```

```
- As a result we have a lot of errors wrapped in errors which makes the code unwieldly and difficult to compose with other code blocks. See image below of incidences of poor composability where we use `firstExceptT` (and sometimes `first`) to wrap errors in other errors.

> ![image](https://github.com/user-attachments/assets/6a50e8f2-9140-4b7e-afa5-ff778d696742)

Reviewer: That's the same image as the first one - I think you meant to use a different one? If that's a code snippet, let's put in a code block.

Reviewer: Ah right, it's a different image. Nevermind.

 
- The ExceptT IO is a known anti-pattern for these reasons and others as per:
  - https://github.com/haskell-effectful/effectful/blob/master/transformers.md
  - https://tech.fpcomplete.com/blog/2016/11/exceptions-best-practices-haskell/

# Decision
- Error handling: All errors will be converted to exceptions that will be caught by a single exception handler at the top level.

> - Top level monad: [RIO](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#t:RIO). See the [ReaderT design pattern](https://tech.fpcomplete.com/blog/2017/06/readert-design-pattern/).

Reviewer: 1. There's no description of a proposed solution. What will be the convention for the  definition exception types? You didn't provide any, so I guess, the goal will be to define an exception near the usage site without [building exceptions hierarchy](https://hackage.haskell.org/package/base-4.21.0.0/docs/Control-Exception.html#t:Exception)?
We should be at least be using `Pretty` for rendering of exceptions, to leverage already existing instances like `Pretty Lovelace` - this needs to be inlcuded in this ADR

2. Why there's suddenly a mention of `ReaderT` design pattern? I don't see how it's related to exception handling.

Author: Proposed exception handling here: https://github.com/IntersectMBO/cardano-cli/pull/1033/commits/989ba8a544277e73548947eb74404f25ddf76d17 

I'll add it to the ADR.

Reviewer: Could you address the second point, please?
